src/Raven.Server/Web/System/AdminDatabasesHandler.cs

```diff
@@ -428,44 +429,67 @@ private async Task WaitForExecutionOnRelevantNodes(JsonOperationContext context,
+
             try
             {
-                foreach (var member in members)
+                using (var cts = CancellationTokenSource.CreateLinkedTokenSource(ServerStore.ServerShutdown))
                 {
-                    var url = clusterTopology.GetUrlFromTag(member);
-                    var requester = ClusterRequestExecutor.CreateForSingleNode(url, ServerStore.Server.Certificate.Certificate);
-                    executors.Add(requester);
-                    waitingTasks.Add(requester.ExecuteAsync(new WaitForRaftIndexCommand(index), context, token: cts.Token));
-                }
+                    cts.CancelAfter(ServerStore.Configuration.Cluster.OperationTimeout.AsTimeSpan);
 
-                while (true)
-                {
-                    var task = await Task.WhenAny(waitingTasks);
-                    if (task == timeoutTask)
-                        throw new TimeoutException($"Waited too long for the raft command (number {index}) to be executed on any of the relevant nodes to this command.");
-                    if (task.IsCompletedSuccessfully)
+                    var waitingTasks = new List<Task>();
+                    List<Task> faultedTasks = null;
+
+                    foreach (var member in members)
+                    {
+                        var url = clusterTopology.GetUrlFromTag(member);
+                        var executor = ClusterRequestExecutor.CreateForSingleNode(url, ServerStore.Server.Certificate.Certificate);
+                        executors.Add(executor);
+                        waitingTasks.Add(executor.ExecuteAsync(new WaitForRaftIndexCommand(index), context, token: cts.Token));
+                    }
+
+                    while (waitingTasks.Count > 0)
                     {
-                        break;
+                        var task = await Task.WhenAny(waitingTasks);
+                        waitingTasks.Remove(task);
+
+                        if (task.IsCompletedSuccessfully)
+                            continue;
+
+                        if (faultedTasks == null)
+                            faultedTasks = new List<Task>();
+
+                        faultedTasks.Add(task);
+                    }
+
+                    if (faultedTasks != null)
+                    {
+                        try
+                        {
+                            // unwrapping exceptions for all of the faulted tasks
+                            await Task.WhenAll(faultedTasks);
+                        }
+                        catch (OperationCanceledException oce)
+                        {
+                            throw new TimeoutException(
+                                $"Waited too long for the raft command (number {index}) to be executed on any of the relevant nodes to this command.", oce);
+                        }
+                        catch (RavenException e) when (e.InnerException is HttpRequestException)
```

Comment:
What if two of them fail with this error, won't we get aggregate exception then?

Reply:
Fixed